cluster-autoscaler/estimator/binpacking_estimator.go

```diff
@@ -68,49 +78,98 @@ func NewBinpackingNodeEstimator(
 // It is assumed that all pods from the given list can fit to nodeTemplate.
 // Returns the number of nodes needed to accommodate all pods from the list.
 func (e *BinpackingNodeEstimator) Estimate(
-	pods []*apiv1.Pod,
+	podsEquivalenceGroups []PodEquivalenceGroup,
 	nodeTemplate *schedulerframework.NodeInfo,
-	nodeGroup cloudprovider.NodeGroup) (int, []*apiv1.Pod) {
+	nodeGroup cloudprovider.NodeGroup,
+) (int, []*apiv1.Pod) {
 
-	e.limiter.StartEstimation(pods, nodeGroup, e.context)
+	e.limiter.StartEstimation(podsEquivalenceGroups, nodeGroup, e.context)
 	defer e.limiter.EndEstimation()
 
-	pods = e.podOrderer.Order(pods, nodeTemplate, nodeGroup)
-
-	newNodeNames := make(map[string]bool)
-	newNodesWithPods := make(map[string]bool)
+	podsEquivalenceGroups = e.podOrderer.Order(podsEquivalenceGroups, nodeTemplate, nodeGroup)
 
+	e.cleanHelperVariables()
 	e.clusterSnapshot.Fork()
 	defer func() {
 		e.clusterSnapshot.Revert()
 	}()
 
-	newNodeNameIndex := 0
 	scheduledPods := []*apiv1.Pod{}
-	lastNodeName := ""
+	for _, podsEquivalenceGroup := range podsEquivalenceGroups {
+		var err error
+		var remainingPods []*apiv1.Pod
+
+		scheduledPods, remainingPods, err = e.tryToScheduleOnExistingNodes(scheduledPods, podsEquivalenceGroup.Pods)
+		if err != nil {
+			klog.Errorf("%v", err)
+			return 0, nil
+		}
 
-	for _, pod := range pods {
-		found := false
+		scheduledPods, err = e.tryToScheduleOnNewNodes(nodeTemplate, scheduledPods, remainingPods)
+		if err != nil {
+			klog.Errorf("%v", err)
+			return 0, nil
+		}
+	}
+
+	if e.estimationAnalyserFunc != nil {
+		e.estimationAnalyserFunc(e.clusterSnapshot, nodeGroup, e.newNodesWithPods)
+	}
+	return len(e.newNodesWithPods), scheduledPods
+}
+
+func (e *BinpackingNodeEstimator) tryToScheduleOnExistingNodes(
+	scheduledPods []*apiv1.Pod,
+	pods []*apiv1.Pod,
+) ([]*apiv1.Pod, []*apiv1.Pod, error) {
+	var index int
+	for index = 0; index < len(pods); index++ {
+		pod := pods[index]
 
+		// Check schedulability on all nodes created during simulation
 		nodeName, err := e.predicateChecker.FitsAnyNodeMatching(e.clusterSnapshot, pod, func(nodeInfo *schedulerframework.NodeInfo) bool {
-			return newNodeNames[nodeInfo.Node().Name]
+			return e.newNodeNames[nodeInfo.Node().Name]
 		})
-		if err == nil {
-			found = true
-			if err := e.clusterSnapshot.AddPod(pod, nodeName); err != nil {
-				klog.Errorf("Error adding pod %v.%v to node %v in ClusterSnapshot; %v", pod.Namespace, pod.Name, nodeName, err)
-				return 0, nil
+		if err != nil {
+			break
+		}
+
+		if err := e.clusterSnapshot.AddPod(pod, nodeName); err != nil {
```

Comment:
This pattern repeats in 3 places: add pod to cluster snapshot, append to `scheduledPods`, update `newNodesWithPods`. Perhaps worth extracting to a function?